ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java

```diff
@@ -0,0 +1,1760 @@
+    CompactionDetails details =
+        new CompactionDetails(state.time.milliseconds() + Time.MsPerSec, Collections.singletonList(lastLogSegmentName));
+    ensureArgumentFailure(details, "Should have failed because compaction range contains offsets still in the journal");
+
+    // compaction contains segments that don't exist
+    details = new CompactionDetails(0,
+        Collections.singletonList(LogSegmentNameHelper.getNextPositionName(lastLogSegmentName)));
+    ensureArgumentFailure(details, "Should have failed because compaction range contains offsets still in the journal");
+  }
+
+  /**
+   * Tests the case where {@link BlobStoreCompactor#resumeCompaction()} is called without any compaction being in
+   * progress.
+   * @throws Exception
+   */
+  @Test
+  public void resumeCompactionWithoutAnyInProgressTest() throws Exception {
+    refreshState(false, true);
+    compactor = getCompactor(state.log, DISK_IO_SCHEDULER);
+    compactor.initialize(state.index);
+    assertFalse("Compaction should not be in progress", CompactionLog.isCompactionInProgress(tempDirStr, STORE_ID));
+    try {
+      compactor.resumeCompaction();
+      fail("Should have failed because there is no compaction in progress");
+    } catch (IllegalStateException e) {
+      // expected. Nothing to do.
+    } finally {
+      compactor.close(0);
+    }
+  }
+
+  /**
+   * A basic test for compaction that selects the first two log segments and compacts them into one log segment.
+   * @throws Exception
+   */
+  @Test
+  public void basicTest() throws Exception {
+    refreshState(false, true);
+    List<String> segmentsUnderCompaction = getLogSegments(0, 2);
+    reduceValidDataSizeInLogSegments(segmentsUnderCompaction, state.log.getSegmentCapacity() - LogSegment.HEADER_SIZE);
+    long deleteReferenceTimeMs = state.time.milliseconds();
+    compactAndVerify(segmentsUnderCompaction, deleteReferenceTimeMs, true);
+  }
+
+  /**
+   * Compacts the whole log (except the last log segment) but without any changes expected i.e all data is valid and is
+   * simply copied over from the old log segments to the new log segments.
+   * @throws Exception
+   */
+  @Test
+  public void compactWholeLogWithNoChangeExpectedTest() throws Exception {
+    long delayBeforeLastLogSegmentWrite = 20 * Time.MsPerSec;
+    refreshState(false, false);
+    // write data until the very last segment is reached
+    long requiredCount = state.log.getCapacityInBytes() / state.log.getSegmentCapacity();
+    // write entries with an expiry time such that no records are actually expired at the time of compaction
+    long expiryTimeMs = getInvalidationTime(requiredCount) + delayBeforeLastLogSegmentWrite;
+    writeDataToMeetRequiredSegmentCount(requiredCount - 1, Collections.singletonList(expiryTimeMs));
+    state.advanceTime(delayBeforeLastLogSegmentWrite);
+    writeDataToMeetRequiredSegmentCount(requiredCount, Collections.singletonList(expiryTimeMs));
```

Comment:
Not sure I understand what you mean. Could you elaborate?

The function takes in as argument, the minimum number of log segments that must exist in the log for it to return i.e. the absolute count - not the incremental count.